### DISCUSSION TITLE
Louvain Clustering over Directed Graphs

### DISCUSSION THREAD

**Discussion by jonathangomesselman** [networkx/networkx#6412] [General]
I have been experimenting with louvain clustering over directed networks and have been noticing some odd (potentially unexpected) behavior. Specifically, I was first surprised to see such a big different between the clustering for a directed graph and its undirected graph equivalent (i.e. treat all edges as undirected and add up edge weights). Although the modularity gain formula is slightly different in the two cases, my impression was that the directed adaptation was meant to fairly closely mirror the undirected case.

Anyways, after doing some digging through the code, I noticed an interesting difference in the `_one_level` function. Specifically, I looked at the different way that the `nbrs` dictionary is created:
```
if is_directed:	
        ...
        # Calculate weights for both in and out neighbours	
        nbrs = {}	
        for u in G:	
            nbrs[u] = defaultdict(float)	
            for _, n, wt in G.out_edges(u, data="weight"):	
                nbrs[u][n] += wt	
            for n, _, wt in G.in_edges(u, data="weight"):	
                nbrs[u][n] += wt	
    else:	
        ...	
        nbrs = {u: {v: data["weight"] for v, data in G[u].items() if v != u} for u in G}
```
What I noticed in particular is that in the undirected case we explicitly exclude `self edges - v != u`. However, we do not do this for the directed case. 

This seems potentially problematic, especially as we merge nodes into "super (community) nodes" during the `_gen_graph` function and add a self-edge with weight representing the weight of intra community edges. Because we include self edges in the `nbrs` the initial `remove cost` for a node with a self edge becomes very large, meaning that we are unlikely to further merge with any other communities. 
```
best_com = node2com[u]
weights2com = _neighbor_weights(nbrs[u], node2com)	
if is_directed:	
                in_degree = in_degrees[u]	
                out_degree = out_degrees[u]	
                Stot_in[best_com] -= in_degree	
                Stot_out[best_com] -= out_degree	
                remove_cost = (	
                    -weights2com[best_com] / m	
                    + resolution	
                    * (out_degree * Stot_in[best_com] + in_degree * Stot_out[best_com])	
                    / m**2	
                )
```
we see that because `u` is in `nbrs[u]`, `weights2com[best_com]` will start as the weight of the self edge, which is likely very high. 

I am curious if this is the intended behavior. Let me know also if I am missing something! Thanks for any feedback.

**Comment by dschult** [comment]
Looking at the history of this module, I suspect an error was introduced in #5713. Before that change both directed and undirected graphs excluded selfloop edges. After it the formula for directed graphs does not exclude selfloops. We should get an example of a directed graph with selfloops (perhaps after a step of building super communities) where we know the proper result and make a test function to check it. We don't have many tests of the directed portion of the code.

I think the obvious solution is to add an `if` clause to each loop that updates `nbrs[u][n]`. The `if` clause would skip selfloops. 

Can you try this, and maybe you have an example or two of directed networks that give unreasonable results now that are improved with your suggested correction. Is there a simple case that exposes the issue?

Thanks!!

**Reply by jonathangomesselman** [reply]
@dschult My initial reaction was as you say to include an `if` clause for each loop that updates `nbrs[u][n]`; however, I wanted to double check that this was not somehow the intended behavior. 

I will try to find some time to look further into your suggestions. The case that I was working on when I noticed this involves a fairly non-trivial graph, so I will work to find some trivial examples. 

Thanks for the quick response here! Glad to see that I wasn't going crazy trying to understand why my experiments weren't making sense.

**Reply by anthonimes** [reply]
Hi @jonathangomesselman I recently came across the same issue and arrived to the same conclusion than you (see PR #6630). 
I will try to update the test to validate this pull request next week!

**Reply by jonathangomesselman** [reply]
Hi @anthonimes @dschult 

Wow I totally forgot about this after somewhat moving on from needing this use case. But yes it seems that you arrived at exactly what I noticed as well. 

I remember when I was looking at this and thinking a bit more about the actual formula for the louvain algorithm it became clear that excluding self-loops is definitely necessary. Otherwise we are incorrectly counting the cost of removing a node from its merged "super-node" community in the next iteration of the algorithm. 

Thanks @anthonimes for creating this PR!
